bpo-30458: Disallow control chars in http URLs.

gpshead opened the pull request:
[bpo-36276](https://bugs.python.org/issue36276) and [bpo-30458](https://bugs.python.org/issue30458):

disallow control characters such as whitespace in urls put into the raw http protocol within http.client.

<!-- issue-number: [bpo-36276](https://bugs.python.org/issue36276) -->
https://bugs.python.org/issue36276
<!-- /issue-number -->

<!-- issue-number: [bpo-30458](https://bugs.python.org/issue30458) -->
https://bugs.python.org/issue30458
<!-- /issue-number -->

gpshead committed:
bpo-14826 bpo-36276: Disallow control chars in http URLs.

Example possible fix for those issues.

tirkarthi commented on Lib/urllib/request.py:
```diff
@@ -350,6 +350,14 @@ def full_url(self):
+        # The same control characters check was adopted by Golang in:
+        #  https://go-review.googlesource.com/c/go/+/159157
+        if (self._full_url.startswith('http') and
+            re.search("[\x00- \x7f-\x9f]", self._full_url)):
```
As per fix in golang this should not include space as CTL and there is a test case where space is counted as a valid. So perhaps the regex could be `[\x00-\x19\x7f-\x9f]` (\x20 for space).

```golang
func isCTL(r rune) bool {
        return r < ' ' || 0x7f <= r && r <= 0x9f
}
```

Sample test case failure with url `http://a b:c d@python.org/` for current regex being used in `_splithost`

```
======================================================================
ERROR: test_userpass_inurl_w_spaces (test.test_urllib.urlopen_HttpTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_urllib.py", line 465, in test_userpass_inurl_w_spaces
    fp = urlopen(url)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_urllib.py", line 47, in urlopen
    return opener.open(url)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/urllib/request.py", line 1735, in open
    urltype, url = _splittype(fullurl)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/urllib/parse.py", line 1003, in _splittype
    raise ValueError("URL can't contain control characters. %r" % (unquoted_url,))
ValueError: URL can't contain control characters. 'http://a b:c d@python.org/'

----------------------------------------------------------------------
Ran 95 tests in 0.650s
```

gpshead replied:
note: `\x1f` is the char before space `\x20` not `\x19` :)

tirkarthi replied:
Sorry, got little confused during conversion `hex(32) # 0x20` and used `\x19` in regex.

gpshead committed:
Move validation to _splittype & correct boundary.

tirkarthi commented on Lib/urllib/parse.py:
```diff
@@ -1000,7 +1002,26 @@ def _splittype(url):
+            global _control_char_re
+            if _control_char_re is None:
+                _control_char_re = re.compile('[\x00-\x1f\x7f-\x9f]')
+            if _control_char_re.search(data):
```
`urlopen` encodes the url https://github.com/python/cpython/blob/750d74fac5c510e39958b3f79641fe54096ee54f/Lib/urllib/request.py#L1730 before calling `_splittype` below and regex match against encoded url might fail. Perhaps check against  `unquote(data)` as a safety check? 

Without `unquote` `urlopen("http://127.0.0.1:1234/?q=HTTP/1.1\r\nHeader: Value")` will not raise a `ValueError` and regex will try to match against `//127.0.0.1:1234/?q=HTTP/1.1%0D%0AHeader:%20Value` here instead of decoded value.

gpshead replied:
Indeed, I noticed.  But realize that your proposed test cases use a _different_ url opening path than urllib.request.urlopen() calls.  We have too many of entry points.  but they all wind up in `http.client` where previous header injection fixes naturally went.

moving this there seems to work out.  our codepath that calls urllib.parse.quote() on the URL is fine, a quoted url is by definition sane and up for the server to handle.  so validating the quoted text will pass.  it's anything that allows raw unquoted urls in that is vulnerable.  such as urllib.request.urlopen() calls or direct http.client.HTTPConnection use.

updated.

also, agreed that we only need to check for ascii also per:

https://github.com/python/cpython/blob/750d74fac5c510e39958b3f79641fe54096ee54f/Lib/http/client.py#L1085

gpshead committed:
Move the check to http.client where it belongs.

gpshead committed:
don't bother checking for chars above ascii.

gpshead committed:
missing f on f-string

gpshead committed:
confirm that repr of the url is in the ValueError.

gpshead committed:
also assert no ' ' in the quoted url

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gpshead committed:
Don't abuse http.client in text_xmlrpc.

Don't abuse http.client to send a bad request via the url parameter.

vstinner commented on Lib/http/client.py:
Would it be possible to mention the first invalid character to help users to debug their code?

vstinner commented on Lib/http/client.py:
```diff
@@ -137,6 +137,15 @@
 _is_illegal_header_value = re.compile(rb'\n(?![ \t])|\r(?![ \t\n])').search
 
+# These characters are not allowed within http URL paths.
+#  https://tools.ietf.org/html/rfc3986#section-3.3
```
Would you mind to describe a little bit what are these characters? It looks like control characters (0..31), space (32) and another control character (127).

tirkarthi replied:
space is not validated. There is `test.test_urllib.urlopen_HttpTests.test_userpass_inurl_w_spaces` that fails if space is validated. golang also didn't check for space.

gpshead replied:
The golang code does not consider space to be a control char (via `< ' '` rather than `<=`). See https://github.com/golang/go/commit/f1d662f34788f4a5f087581d0951cdf4e0f6e708) in their implementation for the CVE.

But space is clearly a forbidden character in a url (% or + encoding required) as the GET line is space delimited.  So I'm keeping out exclusion of space in place.

In our test_userpass_inurl_w_spaces the user@pass has been parsed and removed from the url and turned into an authorization header long before it gets into the http.client code that does the checking so this does not impact usernames or passwords with spaces.

vstinner commented on Lib/http/client.py:
```diff
@@ -137,6 +137,15 @@
 _is_legal_header_name = re.compile(rb'[^:\s][^:\r\n]*').fullmatch
 _is_illegal_header_value = re.compile(rb'\n(?![ \t])|\r(?![ \t\n])').search
 
+# These characters are not allowed within http URL paths.
```
```suggestion
# These characters are not allowed within HTTP URL paths.
```

vstinner reviewed (changes requested):
If we decide to have a single fix for the "various" issues listed in https://bugs.python.org/issue30458 I would prefer to have a much larger test suite. In https://bugs.python.org/issue30458#msg339846 message, I listed around 14 different issues. I expect that we need more tests than just 1 test in test_urllib and 1 test in test_xmlrpc.

For example, I would expect separated tests for space (U+0020), newline characters ('\r' and '\n'), the special case of U+007F, but also for escaped characters using percent.

Can someone please try to collect "test cases" for all issues listed in https://bugs.python.org/issue30458#msg339846 ?

merwok commented on Lib/test/test_xmlrpc.py:
```diff
@@ -943,8 +943,13 @@ def test_unicode_host(self):
+            conn.send('POST /RPC2 HTTP/1.0\r\n'
+                      'Content-Length: 100\r\n\r\n'
+                      f'bye HTTP/1.1\r\n'
+                      'Host: {ADDR}:{PORT}\r\n'
```
It seems like this should be the f-string, not the line before.

Also I don’t get why an existing xmlrpc test is changed by this PR.  It’s testing something else?

tirkarthi replied:
The previous test was sending a partial payload for which it sent headers as part of the URL that was received by the xmlrpc server using httplib. Now that httplib validates against potential header injection in URL it throws exception and hence this was rewritten to use socket.

gpshead replied:
typo from refactoring. fixed.

gpshead committed:
fix misplaced f-str

orsenthil reviewed (approved):
This change, the patch looks good to me. 

I agree with @gpshead 's comment mentioned here: https://github.com/python/cpython/pull/12755#issuecomment-481599611

* Smallest change that addresses the issue, and not changes the public API.


As for @vstinner 's comment, on additional test cases. https://github.com/python/cpython/pull/12755#pullrequestreview-224947379  The bug can still be left in open state for addition of test cases, and this patch can be concentrated on `http/client.py` only.

I will request for @vstinner 's approval as well.

gpshead committed:
Address code review comments.

Adds some comments and expands an error message per vstinner's PR
review.

(authored and pushed from 32,000ft on the plane to CLE for PyCon US)

gpshead committed:
fix indentation

tirkarthi reviewed (approved):
LGTM. Getting this to testing will help.

Thanks @gpshead for the PR :)

This PR is also adopted in urllib3 : https://github.com/urllib3/urllib3/pull/1591

gpshead committed:
Expand the tests to cover every char.

all rejected chars are explicitly tested now.
